hoomd/array.py

```diff
@@ -0,0 +1,652 @@
+            tuple.
+        *args (Any): Required position arguments for the functors.
+        allow_exceptions (bool, optional): A key word only arugment that
+            defaults to False. When True exceptions are ignored when setting
+            class methods, and the method raising the error is skipped.
+        **kwargs (Any): Required key word arugments for the functors.
+    """
+    class _WrapClass(type):
+        def __new__(cls, name, bases, class_dict):
+            for methods, functor in methods_wrap_func_list:
+                for method in methods:
+                    try:
+                        class_dict[method] = functor(method, *args, **kwargs)
+                    except Exception as err:
+                        if allow_exceptions:
+                            continue
+                        else:
+                            raise err from None
+
+            return super().__new__(cls, name, bases, class_dict)
+
+    return _WrapClass
+
+
+"""Various list of NumPy ndarray functions.
+
+We separate them out by the kind of wrapping they need. We have to distinguish
+between functions that return a new array, functions that return the same array,
+and functions that return a new array with the same underlying data.
+"""
+
+# Functions that return a new array and wrapper
+
+
+def _op_wrap(method, cls=np.ndarray):
+    func = getattr(cls, method)
+
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+        arr = self._coerce_to_ndarray()
+        return func(arr, *args, **kwargs)
+
+    return wrapped
+
+
+_ndarray_ops_ = ([
+    # Comparison
+    '__lt__', '__le__', '__gt__', '__ge__', '__eq__', '__ne__', '__bool__',
+    # Unary
+    '__neg__', '__pos__', '__abs__', '__invert__',
+    # Arithmetic
+    '__add__', '__sub__', '__mul__', '__truediv__', '__floordiv__', '__mod__',
+    '__divmod__', '__pow__',
+    # Bitwise
+    '__lshift__', '__rshift__', '__and__', '__or__', '__xor__',
+    # Matrix
+    '__matmul__',
+], _op_wrap)
+
+# Magic methods that never return an array to the same underlying buffer
+_magic_wrap = _op_wrap
+
+_ndarray_magic_safe_ = ([
+    # Copy
+    '__copy__', '__deepcopy__',
+    # Pickling
+    '__reduce__', '__setstate__',
+    # Container based
+    '__len__', '__setitem__', '__contains__',
+    # Conversion
+    '__int__', '__float__', '__complex__'
+], _magic_wrap)
+
+# Magic methods that may return an array pointing to the same buffer
+
+
+def _magic_wrap_with_check(method, cls=np.ndarray):
+    func = getattr(cls, method)
+
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+        arr = self._coerce_to_ndarray()
+        rtn = func(arr, *args, **kwargs)
+        if isinstance(rtn, np.ndarray) and np.may_share_memory(rtn, arr):
+            return self.__class__(rtn, self._callback, self.read_only)
+        else:
+            return rtn
+
+    return wrapped
+
+
+_ndarray_magic_unsafe_ = ([
```

Comment:
Is this "unsafe" because we don't have any guarantee whether the data is the same?

Reply:
It isn't unsafe for us, it is potentially unsafe (i.e. we need to make sure we don't leak a memory buffer, if the returned objects uses the same memory).